### PR TITLE
Feat-Implement httpcache middleware for GitHub API

### PR DESCRIPTION
**Pull request by naveensrinivasan**
* **Please check if the PR fulfills these requirements**
- [x] Tests for the changes have been added (for bug fixes/features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)
Feature - Caching

* **What is the current behavior?** (You can also link to an open issue here)
#80 #202 

* **What is the new behavior (if this is a feature change)?**
The GitHub API supports conditional requests
https://docs.github.com/en/rest/overview/resources-in-the-rest-api#conditional-requests

https://github.com/google/go-github supports Conditional requests
https://github.com/google/go-github#conditional-requests

As we are scaling more and more projects this would add a lot of value.

Initial run fetches information using `httpcache` as a middleware,
which caches the HTTP response initially in a large disk (PVC),
probably move to Redis later as a cache instead of disk.

Subsequent `cron runs` will utilize the `httpcache` for checking content modification and
load it from the cache if it isn't modified, which reduces the hitting the
Rate Limit of the GitHub API.

Also fixed the golang-ci warnings.

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)
None

* **Other information**:

Subsequent cache runs on 50 repositories takes about 18 minutes with `3` GitHub tokens

```
❯ time ./cron.sh
github.com/18F/identity-idp
github.com/1N3/Sn1per
github.com/24pullrequests/24pullrequests
github.com/AElfProject/AElf
github.com/Activiti/Activiti
github.com/AdguardTeam/AdguardFilters
github.com/AlchemyCMS/alchemy_cms
github.com/AliasIO/wappalyzer
github.com/Amanieu/parking_lot
github.com/abonas/kubeclient
github.com/abpframework/abp
github.com/acassen/keepalived
github.com/acmesh-official/acme.sh
github.com/actions/runner
github.com/activeadmin/activeadmin
github.com/activemerchant/active_merchant
github.com/activerecord-hackery/ransack
github.com/activescaffold/active_scaffold
github.com/actix/actix
github.com/actix/actix-extras
github.com/actix/actix-net
github.com/actix/actix-web
github.com/adafruit/Adafruit_CircuitPython_Bundle
github.com/adempiere/adempiere
github.com/afragen/github-updater
github.com/aframevr/aframe
github.com/ai/autoprefixer-rails
github.com/aio-libs/aiohttp
github.com/airbnb/javascript
github.com/airbrake/airbrake
github.com/akaunting/akaunting
github.com/akeneo/pim-community-dev
github.com/akka/akka
github.com/akka/akka-http
github.com/akkadotnet/akka.net
github.com/alacritty/alacritty
github.com/alanmcgovern/monotorrent
github.com/alexcrichton/cc-rs
github.com/alexcrichton/curl-rust
github.com/alexcrichton/proc-macro2
github.com/alexcrichton/toml-rs
github.com/alexreisner/geocoder
github.com/alibaba/dragonwell8
github.com/alibaba/druid
github.com/alibaba/fastjson
github.com/alibaba/nacos
github.com/allenai/allennlp
github.com/allinurl/goaccess
github.com/alphagov/whitehall
github.com/alpinelinux/aports
./cron.sh  46.68s user 19.95s system 6% cpu 18:26.70 total
```

### Folder size 

```
du -sh ./cache
801M	
```

### Files in the folder

```
find ./cache -type f | wc -l
10347
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -7,10 +7,11 @@ require (
 	github.com/golang/protobuf v1.4.3 // indirect
 	github.com/google/go-github/v32 v32.1.0
 	github.com/kr/text v0.2.0 // indirect
+	github.com/naveensrinivasan/httpcache v1.2.1
```
Comment (naveensrinivasan) [inline go.mod:10]: This is a fork of https://github.com/gregjones/httpcache/pull/104 this PR.

File: .gitignore [hunk 2]
```diff
@@ -23,3 +23,6 @@ results.json
 
 # tools 
 bin
+
+#temp
```
Comment (inferno-chromium) [inline .gitignore:27]: remove this.

File: README.md [hunk 3]
```diff
@@ -153,6 +152,20 @@ Signed-Releases: Fail 0
 Signed-Tags: Fail 10
 ```
 
+### Caching
+
+Scorecard uses `httpcache` with <https://docs.github.com/en/rest/overview/resources-in-the-rest-api#conditional-requests> for caching httpresponse. The default cache is in-memory.
```
Comment (dlorenc) [inline README.md:157]: Maybe link to the etags stuff in GitHub - the real benefit is avoiding the API quota

File: roundtripper/roundtripper.go [hunk 4]
```diff
@@ -90,22 +90,37 @@ func NewTransport(ctx context.Context, logger *zap.SugaredLogger) http.RoundTrip
...
+	// uses memory cache
+	c := cache.NewTransport(cache.NewMemoryCache())
+	c.Transport = rateLimit
+	return c
+}
+
+// shouldUseDiskCache checks the env variables USE_DISK_CACHE and DISK_CACHE_PATH to determine if
+// disk should be used for caching.
+func shouldUseDiskCache() (string, bool) {
+	if isDiskCache := os.Getenv(UseDiskCache); isDiskCache != "" {
```
Comment (dlorenc) [inline roundtripper/roundtripper.go:109]: nit: I think you can avoid this if statement and go straight into ParseBool since "" parses as false.

**Review by dlorenc — COMMENTED** [review]
A couple tiny nits, LGTM!

File: README.md [hunk 5]
```diff
@@ -153,6 +152,20 @@ Signed-Releases: Fail 0
...
+
+#### Disk Cache
+
+Scorecard results can be cached into a disk for increasing throughput for subsequent runs.
+
+To use disk cache two env variables have to be set `USE_DISK_CACHE=true` and `DISK_CACHE_PATH=./cache`.
+
+There are not TTL on cache.
```
Comment (inferno-chromium) [inline README.md:165]: "There is no"

File: roundtripper/roundtripper.go [hunk 6]
```diff
@@ -90,22 +90,37 @@ func NewTransport(ctx context.Context, logger *zap.SugaredLogger) http.RoundTrip
...
+		if result, err := strconv.ParseBool(isDiskCache); err == nil && result {
+			if cachePath := os.Getenv(DiskCachePath); cachePath != "" {
+				return cachePath, true
+			}
+		}
+	}
+	return "", false
```
Comment (inferno-chromium) [inline roundtripper/roundtripper.go:116]: nit: maybe nil instead of ""

File: roundtripper/roundtripper_test.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	if e != nil {
+		t.Errorf(e.Error())
+	}
+}
+
+func Test_shouldUseDiskCache(t *testing.T) {
+	t.Parallel()
+	tests := []struct {
+		name  string
+		want  string
```
Comment (inferno-chromium) [inline roundtripper/roundtripper_test.go:32]: s/want/diskCachePath
s/want1/useDiskCache

**Review by inferno-chromium — APPROVED** [review]
Thanks a lot! This is very exciting.

File: roundtripper/roundtripper.go (hunk 6 above)
Comment (naveensrinivasan) [reply roundtripper/roundtripper.go:116]: can't do nil for a string in `go`

**Pull request merged**